### PR TITLE
More abstraction of crypto protocol implementation

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

**Comment by djc** [comment]
The logging for errors is a bit worse, but I couldn't think of a better way. Any ideas?

File: quinn-proto/src/crypto.rs [hunk 1]
```diff
@@ -64,8 +64,14 @@ impl CryptoSession for TlsSession {
         }
     }
 
-    fn read_handshake(&mut self, buf: &[u8]) -> Result<(), TLSError> {
-        self.read_hs(buf)
+    fn read_handshake(&mut self, buf: &[u8]) -> Result<(), TransportError> {
+        self.read_hs(buf).map_err(|_| {
+            if let Some(code) = self.error_code() {
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: I think we can remove `error_code` from the trait entirely with this change.

Comment (djc) [reply quinn-proto/src/crypto.rs]: Right! Fixed.

**Comment by djc** [comment]
Ready for another round of review, I think.

File: quinn-proto/src/crypto.rs [hunk 2]
```diff
@@ -64,8 +59,14 @@ impl CryptoSession for TlsSession {
         }
     }
 
-    fn read_handshake(&mut self, buf: &[u8]) -> Result<(), TLSError> {
-        self.read_hs(buf)
+    fn read_handshake(&mut self, buf: &[u8]) -> Result<(), TransportError> {
+        self.read_hs(buf).map_err(|e| {
+            if let Some(alert) = self.get_alert() {
+                TransportError::crypto(alert.get_u8(), format!("TLS error: {}", e))
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]:
```suggestion
                TransportError::crypto(alert.get_u8(), e.to_string())
```
TLS-ness is implied by the error code.

File: quinn-proto/src/connection.rs [hunk 3]
```diff
@@ -906,12 +908,8 @@ impl Connection {
             }
             trace!(self.log, "read {} TLS bytes", n);
             if let Err(e) = self.tls.read_handshake(&buf[..n]) {
-                debug!(self.log, "TLS error: {}", e);
-                return Err(if let Some(alert) = self.tls.alert() {
-                    TransportError::crypto(alert)
-                } else {
-                    TransportError::PROTOCOL_VIOLATION("TLS error")
-                });
+                debug!(self.log, "reading handshake bytes resulted in {}", e);
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]: `TransportError`s are already logged in `handle_packet`, and we're now encoding all information into them, so this is redundant. Should be able to do away with the `if let` entirely.

Comment (djc) [reply quinn-proto/src/connection.rs]: Fixed.

**Pull request merged**